orchagent/neighorch.cpp

```diff
@@ -784,39 +824,12 @@ bool NeighOrch::removeNeighbor(const NeighborEntry &neighborEntry, bool disable)
         neighbor_entry.switch_id = gSwitchId;
         copy(neighbor_entry.ip_address, ip_address);
 
-        sai_object_id_t next_hop_id = m_syncdNextHops[nexthop].next_hop_id;
-        status = sai_next_hop_api->remove_next_hop(next_hop_id);
-        if (status != SAI_STATUS_SUCCESS)
-        {
-            /* When next hop is not found, we continue to remove neighbor entry. */
-            if (status == SAI_STATUS_ITEM_NOT_FOUND)
-            {
-                SWSS_LOG_ERROR("Failed to locate next hop %s on %s, rv:%d",
-                               ip_address.to_string().c_str(), alias.c_str(), status);
-            }
-            else
-            {
-                SWSS_LOG_ERROR("Failed to remove next hop %s on %s, rv:%d",
-                               ip_address.to_string().c_str(), alias.c_str(), status);
-                return handleSaiRemoveStatus(SAI_API_NEXT_HOP, status);
-            }
-        }
-
-        if (status != SAI_STATUS_ITEM_NOT_FOUND)
+        /* If we can't remove the next hop, return false. */
+        if (!removeNextHop(nexthop))
```

Comment:
The flow is slightly changed here. Previously we remove this nexthop from the map only after neighbor is successfully removed. What is the motivation?

Reply:
@TACappleman Tom made these changes and can address your question.

Reply:
The main part of this change is to move the actual deletion of the next hop over the SAI into the removeNextHop (so that when e.g. routeorch calls this method it does what you'd expect it to). It then makes sense for the deletion of the next hop from the map to happen at the point that it no longer exists on the SAI, and so the small functional change as a result of this set of changes seems reasonable. (Otherwise you could delete the next hop and fail to delete the neighbor, and so not record that the next hop has been deleted).

Reply:
Please change it to original flow and have only the changes for mpls addition. Else its difficult to understand what is optimized here.

Reply:
@prsunny  You need to re-examine the original code.  You will see there is NextHop clean-up code in ::removeNeighbor instead of in ::removeNextHop.  This only worked because for IP there is a 1-1 mapping between NextHop and Neighbor, so the NextHop was only removed when the Neighbor was also being removed.  With MPLS NHs, this is no longer true.  Now there is n-1 mapping between NH and Neighbor, so NextHop cleanup *must* occur within the context of ::removeNextHop to ascertain that NH is truly cleaned up.  ::removeNeighbor can (and does) call :;removeNextHop to achieve its previous functionality.

Reply:
previous flow was 
1. SAI Remove nexthop 2. SAI remove neighbor 3. Neighbor map remove 4. Nexthop Map remove

new flow:
1. SAI Remove nexthop 2  Nexthop Map remove 3. SAI remove neighbor 4. Neighbor map remove

How does mpls n-1:1 mapping between nh and neighbor cover new flow ?

Reply:
@abdosi With MPLS NHs, you can no longer assume that Neighbor is removed when NextHop is removed.  I am puzzled as to why this flow does not remove Neighbor/Nexthop in the reverse order that Neighbor/Nexthop is created.  Can you explain why addNextHop() has SAI API calls, but removeNextHop() does not?   There seems to be some (possible unintentional) asymmetry here that needs to be explained.

Reply:
@prsunny Per the discussion, I have reverted NeighOrch::removeNextHop and added separate NeighOrch::removeMplsNextHop.